databuilder/query_language.py

```diff
@@ -32,13 +32,21 @@ def compile(self) -> dict[str, Node]:  # noqa A003
 
 
 class PatientTable:
-    pass
+    def __init__(self):
+        self.qm_node = SelectPatientTable(name=self.__name__)
 
 
 class Series:
-    def __init__(self, qm_node):
+    def __init__(self, qm_node=None):
         self.qm_node = qm_node
 
+    def __set_name__(self, owner, name):
+        self.__name__ = name
```

Comment:
I missed `__set_name__` on my first readthrough -- what's it for?

Reply:
Oh, I see, it's so we can do `self.__name__` in the descriptor.

Reply:
Yes, good question. This is how a descriptor knows the name it has been assigned by the containing class. I expected the name to be passed to `Series.__get__`, but it isn't, as far as I can tell from the documentation and some `pdb`ing. I followed the [Customized names][1] section of the Descriptor HowTo Guide, which uses this to give public and private names.

[1]: https://docs.python.org/3.9/howto/descriptor.html#customized-names

Reply:
Ah, so, the need for `__set_name__` makes more sense now: the right-hand side of an assignment is executed before the corresponding variable exists, so given:

```python
class Patients(PatientTable):
    patient_id = IdSeries()
    ...
```

`IdSeries` is instantiated before `Patients.patient_id` exists. Prior to the introduction of `__set_name__` in Python 3.6, passing the variable's name to the descriptor necessitated some duplication:

```python
class Patients(PatientTable):
    patient_id = IdSeries("patient_id")
    ...
```

Or a workaround involving an auto-generated name. [_Attribute Descriptors_](https://learning.oreilly.com/library/view/fluent-python/9781491946237/ch20.html) in _Fluent Python_ gives an example.

[This gist](https://gist.github.com/iaindillingham/71aa6d207a0370d236520d4c313d7698) demonstrates the order of execution.

Reply:
This all looks correct to me.